### ISSUE TITLE
Question about forward pass for unlabel samples in training process.

### ISSUE THREAD

**Issue by DingYuan0118** [YU1ut/MixMatch-pytorch#43]

**Comment by DingYuan0118** [comment]
In [`train.py` line 228-235](https://github.com/YU1ut/MixMatch-pytorch/blob/master/train.py#L228-235) , there is
```
with torch.no_grad():
    # compute guessed labels of unlabel samples
    outputs_u = model(inputs_u)
    outputs_u2 = model(inputs_u2)
    p = (torch.softmax(outputs_u, dim=1) + torch.softmax(outputs_u2, dim=1)) / 2
    pt = p**(1/args.T)
    targets_u = pt / pt.sum(dim=1, keepdim=True)
    targets_u = targets_u.detach()
```
At this time, the **training mode** for the model is `True` means `model.training = True`. The Batchnorm layer will track the mean and std in the current unlabeled sample batch and update the `running_mean` and `running_var` of the BN layer. Should we set `model.eval()` here like
```
with torch.no_grad():
    # compute guessed labels of unlabel samples
    model.eval()
    outputs_u = model(inputs_u)
    outputs_u2 = model(inputs_u2)
    p = (torch.softmax(outputs_u, dim=1) + torch.softmax(outputs_u2, dim=1)) / 2
    pt = p**(1/args.T)
    targets_u = pt / pt.sum(dim=1, keepdim=True)
    targets_u = targets_u.detach()
    model.train()
```
to avoid repeated updates in the case of unlabeled samples?

**Comment by DingYuan0118** [comment]
I have tried this method, but the result failed and the training loss is unstable. I don`t know what's wrong with it. Could you please give me some explanation?

**Comment by DingYuan0118** [comment]
I have found the answer after running a lot of ablation experiments.

the `model.eval()` changes the way the `BatchNorm2d` layer behaves. When in the `eval()` mode, BN layer will use the `running_mean` and `running_var` to compute the result, which is improper since the `running_mean` and `running_var` is initialized as 0 and 1 when the training process starts. So if you want to avoid repeated updates for BN statistics as I describe above, you should do things like this:
```py
import torch.nn as nn
def freeze_bn(m):
    if isinstance(m, nn,BatchNorm2d):
        m.track_running_stats=False

def activate_bn(m):
    if isinstance(m, nn.BatchNorm2d):
        m.track_running_stats=True

with torch.no_grad():
    # compute guessed labels of unlabel samples
    model.apply(freeze_bn)
    outputs_u = model(inputs_u)
    outputs_u2 = model(inputs_u2)
    p = (torch.softmax(outputs_u, dim=1) + torch.softmax(outputs_u2, dim=1)) / 2
    pt = p**(1/args.T)
    targets_u = pt / pt.sum(dim=1, keepdim=True)
    targets_u = targets_u.detach()
    model.apply(activate_bn)
```

This will produce the right result.
